In [1]:
import pandas as pd
import numpy as np
from scipy.stats import pointbiserialr

file_path = 'diabetic_data.csv'  
df = pd.read_csv(file_path)

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 50 columns):
 #   Column                    Non-Null Count   Dtype 
---  ------                    --------------   ----- 
 0   encounter_id              101766 non-null  int64 
 1   patient_nbr               101766 non-null  int64 
 2   race                      101766 non-null  object
 3   gender                    101766 non-null  object
 4   age                       101766 non-null  object
 5   weight                    101766 non-null  object
 6   admission_type_id         101766 non-null  int64 
 7   discharge_disposition_id  101766 non-null  int64 
 8   admission_source_id       101766 non-null  int64 
 9   time_in_hospital          101766 non-null  int64 
 10  payer_code                101766 non-null  object
 11  medical_specialty         101766 non-null  object
 12  num_lab_procedures        101766 non-null  int64 
 13  num_procedures            101766 non-null  int64 
 14  num_

In [3]:
df.describe()

encounter_id   patient_nbr  admission_type_id  \
count  1.017660e+05  1.017660e+05      101766.000000   
mean   1.652016e+08  5.433040e+07           2.024006   
std    1.026403e+08  3.869636e+07           1.445403   
min    1.252200e+04  1.350000e+02           1.000000   
25%    8.496119e+07  2.341322e+07           1.000000   
50%    1.523890e+08  4.550514e+07           1.000000   
75%    2.302709e+08  8.754595e+07           3.000000   
max    4.438672e+08  1.895026e+08           8.000000   

       discharge_disposition_id  admission_source_id  time_in_hospital  \
count             101766.000000        101766.000000     101766.000000   
mean                   3.715642             5.754437          4.395987   
std                    5.280166             4.064081          2.985108   
min                    1.000000             1.000000          1.000000   
25%                    1.000000             1.000000          2.000000   
50%                    1.000000             7.000000          4.000000   
75%                    4.000000             7.000000          6.000000   
max                   28.000000            25.000000         14.000000   

       num_lab_procedures  num_procedures  num_medications  number_outpatient  \
count       101766.000000   101766.000000    101766.000000      101766.000000   
mean            43.095641        1.339730        16.021844           0.369357   
std             19.674362        1.705807         8.127566           1.267265   
min              1.000000        0.000000         1.000000           0.000000   
25%             31.000000        0.000000        10.000000           0.000000   
50%             44.000000        1.000000        15.000000           0.000000   
75%             57.000000        2.000000        20.000000           0.000000   
max            132.000000        6.000000        81.000000          42.000000   

       number_emergency  number_inpatient  number_diagnoses  
count     101766.000000     101766.000000     101766.000000  
mean           0.197836          0.635566          7.422607  
std            0.930472          1.262863          1.933600  
min            0.000000          0.000000          1.000000  
25%            0.000000          0.000000          6.000000  
50%            0.000000          0.000000          8.000000  
75%            0.000000          1.000000          9.000000  
max           76.000000         21.000000         16.000000

In [4]:
df.shape

(101766, 50)

In [5]:
df.columns

Index(['encounter_id', 'patient_nbr', 'race', 'gender', 'age', 'weight',
       'admission_type_id', 'discharge_disposition_id', 'admission_source_id',
       'time_in_hospital', 'payer_code', 'medical_specialty',
       'num_lab_procedures', 'num_procedures', 'num_medications',
       'number_outpatient', 'number_emergency', 'number_inpatient', 'diag_1',
       'diag_2', 'diag_3', 'number_diagnoses', 'max_glu_serum', 'A1Cresult',
       'metformin', 'repaglinide', 'nateglinide', 'chlorpropamide',
       'glimepiride', 'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide',
       'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',
       'tolazamide', 'examide', 'citoglipton', 'insulin',
       'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted'],
      dtype='object')

In [6]:
df.head()

encounter_id  patient_nbr             race  gender      age weight  \
0       2278392      8222157        Caucasian  Female   [0-10)      ?   
1        149190     55629189        Caucasian  Female  [10-20)      ?   
2         64410     86047875  AfricanAmerican  Female  [20-30)      ?   
3        500364     82442376        Caucasian    Male  [30-40)      ?   
4         16680     42519267        Caucasian    Male  [40-50)      ?   

   admission_type_id  discharge_disposition_id  admission_source_id  \
0                  6                        25                    1   
1                  1                         1                    7   
2                  1                         1                    7   
3                  1                         1                    7   
4                  1                         1                    7   

   time_in_hospital  ... citoglipton insulin  glyburide-metformin  \
0                 1  ...          No      No                   No   
1                 3  ...          No      Up                   No   
2                 2  ...          No      No                   No   
3                 2  ...          No      Up                   No   
4                 1  ...          No  Steady                   No   

   glipizide-metformin  glimepiride-pioglitazone  metformin-rosiglitazone  \
0                   No                        No                       No   
1                   No                        No                       No   
2                   No                        No                       No   
3                   No                        No                       No   
4                   No                        No                       No   

   metformin-pioglitazone  change diabetesMed readmitted  
0                      No      No          No         NO  
1                      No      Ch         Yes        >30  
2                      No      No         Yes         NO  
3                      No      Ch         Yes         NO  
4                      No      Ch         Yes         NO  

[5 rows x 50 columns]

In [7]:
df.replace('?', np.nan, inplace=True)

In [8]:
df.duplicated().sum()

0

In [9]:
df.isnull().sum()

encounter_id                    0
patient_nbr                     0
race                         2273
gender                          0
age                             0
weight                      98569
admission_type_id               0
discharge_disposition_id        0
admission_source_id             0
time_in_hospital                0
payer_code                  40256
medical_specialty           49949
num_lab_procedures              0
num_procedures                  0
num_medications                 0
number_outpatient               0
number_emergency                0
number_inpatient                0
diag_1                         21
diag_2                        358
diag_3                       1423
number_diagnoses                0
max_glu_serum               96420
A1Cresult                   84748
metformin                       0
repaglinide                     0
nateglinide                     0
chlorpropamide                  0
glimepiride                     0
acetohexamide 

In [10]:
for col in df.columns:
    if df[col].isnull().sum() > 0:
        if df[col].dtype == 'object':
            df[col].fillna(df[col].mode()[0], inplace=True)
        else:
            df[col].fillna(df[col].median(), inplace=True) 

C:\Users\DLK Groups\AppData\Local\Temp\ipykernel_6872\3468761712.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].fillna(df[col].mode()[0], inplace=True)


In [11]:
df.isnull().sum()

encounter_id                0
patient_nbr                 0
race                        0
gender                      0
age                         0
weight                      0
admission_type_id           0
discharge_disposition_id    0
admission_source_id         0
time_in_hospital            0
payer_code                  0
medical_specialty           0
num_lab_procedures          0
num_procedures              0
num_medications             0
number_outpatient           0
number_emergency            0
number_inpatient            0
diag_1                      0
diag_2                      0
diag_3                      0
number_diagnoses            0
max_glu_serum               0
A1Cresult                   0
metformin                   0
repaglinide                 0
nateglinide                 0
chlorpropamide              0
glimepiride                 0
acetohexamide               0
glipizide                   0
glyburide                   0
tolbutamide                 0
pioglitazo

In [12]:
df['encounter_id'].value_counts()

encounter_id
2278392      1
190792044    1
190790070    1
190789722    1
190786806    1
            ..
106665324    1
106657776    1
106644876    1
106644474    1
443867222    1
Name: count, Length: 101766, dtype: int64

In [13]:
df['patient_nbr'].value_counts()

patient_nbr
88785891     40
43140906     28
1660293      23
88227540     23
23199021     23
             ..
11005362      1
98252496      1
1019673       1
13396320      1
175429310     1
Name: count, Length: 71518, dtype: int64

In [14]:
df['race'].value_counts()

race
Caucasian          78372
AfricanAmerican    19210
Hispanic            2037
Other               1506
Asian                641
Name: count, dtype: int64

In [15]:
df['gender'].value_counts()

gender
Female             54708
Male               47055
Unknown/Invalid        3
Name: count, dtype: int64

In [16]:
df['age'].value_counts()

age
[70-80)     26068
[60-70)     22483
[50-60)     17256
[80-90)     17197
[40-50)      9685
[30-40)      3775
[90-100)     2793
[20-30)      1657
[10-20)       691
[0-10)        161
Name: count, dtype: int64

In [17]:
def convert_age_range_to_midpoint(age_range):
    try:
        low, high = map(int, age_range.strip('[]()').split('-'))
        return (low + high) / 2
    except:
        return np.nan
df['age'] = df['age'].apply(convert_age_range_to_midpoint)

In [18]:
df['age'].value_counts()

age
75.0    26068
65.0    22483
55.0    17256
85.0    17197
45.0     9685
35.0     3775
95.0     2793
25.0     1657
15.0      691
5.0       161
Name: count, dtype: int64

In [19]:
df['weight'].value_counts()

weight
[75-100)     99905
[50-75)        897
[100-125)      625
[125-150)      145
[25-50)         97
[0-25)          48
[150-175)       35
[175-200)       11
>200             3
Name: count, dtype: int64

In [20]:
def convert_weight_range_to_midpoint(weight):
    if isinstance(weight, str) and weight.startswith('>'):
        return float(weight[1:]) + 10  # You can adjust +10 based on your logic
    try:
        low, high = map(int, weight.strip('[]()').split('-'))
        return (low + high) / 2
    except:
        return np.nan
df['weight'] = df['weight'].apply(convert_weight_range_to_midpoint)

In [21]:
df['weight'].value_counts()

weight
87.5     99905
62.5       897
112.5      625
137.5      145
37.5        97
12.5        48
162.5       35
187.5       11
210.0        3
Name: count, dtype: int64

In [22]:
df['admission_type_id'].value_counts()

admission_type_id
1    53990
3    18869
2    18480
6     5291
5     4785
8      320
7       21
4       10
Name: count, dtype: int64

In [23]:
df['discharge_disposition_id'].value_counts()

discharge_disposition_id
1     60234
3     13954
6     12902
18     3691
2      2128
22     1993
11     1642
5      1184
25      989
4       815
7       623
23      412
13      399
14      372
28      139
8       108
15       63
24       48
9        21
17       14
16       11
19        8
10        6
27        5
12        3
20        2
Name: count, dtype: int64

In [24]:
df['admission_source_id'].value_counts()

admission_source_id
7     57494
1     29565
17     6781
4      3187
6      2264
2      1104
5       855
3       187
20      161
9       125
8        16
22       12
10        8
14        2
11        2
25        2
13        1
Name: count, dtype: int64

In [25]:
df['time_in_hospital'].value_counts()

time_in_hospital
3     17756
2     17224
1     14208
4     13924
5      9966
6      7539
7      5859
8      4391
9      3002
10     2342
11     1855
12     1448
13     1210
14     1042
Name: count, dtype: int64

In [26]:
df['payer_code'].value_counts()

payer_code
MC    72695
HM     6274
SP     5007
BC     4655
MD     3532
CP     2533
UN     2448
CM     1937
OG     1033
PO      592
DM      549
CH      146
WC      135
OT       95
MP       79
SI       55
FR        1
Name: count, dtype: int64

In [27]:
df['medical_specialty'].value_counts()

medical_specialty
InternalMedicine                    64584
Emergency/Trauma                     7565
Family/GeneralPractice               7440
Cardiology                           5352
Surgery-General                      3099
                                    ...  
Perinatology                            1
Neurophysiology                         1
Psychiatry-Addictive                    1
Pediatrics-InfectiousDiseases           1
Surgery-PlasticwithinHeadandNeck        1
Name: count, Length: 72, dtype: int64

In [28]:
df['num_lab_procedures'].value_counts()

num_lab_procedures
1      3208
43     2804
44     2496
45     2376
38     2213
       ... 
120       1
132       1
121       1
126       1
118       1
Name: count, Length: 118, dtype: int64

In [29]:
df['num_procedures'].value_counts()

num_procedures
0    46652
1    20742
2    12717
3     9443
6     4954
4     4180
5     3078
Name: count, dtype: int64

In [30]:
df['num_medications'].value_counts()

num_medications
13    6086
12    6004
11    5795
15    5792
14    5707
      ... 
70       2
75       2
81       1
79       1
74       1
Name: count, Length: 75, dtype: int64

In [31]:
df['number_outpatient'].value_counts()

number_outpatient
0     85027
1      8547
2      3594
3      2042
4      1099
5       533
6       303
7       155
8        98
9        83
10       57
11       42
13       31
12       30
14       28
15       20
16       15
17        8
21        7
20        7
18        5
22        5
19        3
27        3
24        3
26        2
23        2
25        2
33        2
35        2
36        2
29        2
34        1
39        1
42        1
28        1
37        1
38        1
40        1
Name: count, dtype: int64

In [32]:
df['number_emergency'].value_counts()

number_emergency
0     90383
1      7677
2      2042
3       725
4       374
5       192
6        94
7        73
8        50
10       34
9        33
11       23
13       12
12       10
22        6
16        5
18        5
19        4
20        4
15        3
14        3
25        2
21        2
28        1
42        1
46        1
76        1
37        1
64        1
63        1
54        1
24        1
29        1
Name: count, dtype: int64

In [33]:
df['number_inpatient'].value_counts()

number_inpatient
0     67630
1     19521
2      7566
3      3411
4      1622
5       812
6       480
7       268
8       151
9       111
10       61
11       49
12       34
13       20
14       10
15        9
16        6
19        2
17        1
21        1
18        1
Name: count, dtype: int64

In [34]:
df['diag_1'].value_counts()

diag_1
428    6883
414    6581
786    4016
410    3614
486    3508
       ... 
817       1
61        1
148       1
870       1
V51       1
Name: count, Length: 716, dtype: int64

In [35]:
df['diag_2'].value_counts()

diag_2
276     7110
428     6662
250     6071
427     5036
401     3736
        ... 
232        1
908        1
52         1
E817       1
927        1
Name: count, Length: 748, dtype: int64

In [36]:
df['diag_3'].value_counts()

diag_3
250     12978
401      8289
276      5175
428      4577
427      3955
        ...  
657         1
684         1
603         1
E826        1
971         1
Name: count, Length: 789, dtype: int64

In [37]:
df['number_diagnoses'].value_counts()

number_diagnoses
9     49474
5     11393
8     10616
7     10393
6     10161
4      5537
3      2835
2      1023
1       219
16       45
10       17
13       16
11       11
15       10
12        9
14        7
Name: count, dtype: int64

In [38]:
df['max_glu_serum'].value_counts()

max_glu_serum
Norm    99017
>200     1485
>300     1264
Name: count, dtype: int64

In [39]:
def label_max_glu_serum(value):
    if pd.isnull(value) or value == 'None':
        return 'None'
    elif value == 'Norm':
        return 'Normal'
    elif value == '>200':
        return 'Greater than 200'
    elif value == '>300':
        return 'Greater than 300'
    else:
        return 'Unknown'
df['max_glu_serum'] = df['max_glu_serum'].apply(label_max_glu_serum)

In [40]:
df['max_glu_serum'].value_counts()

max_glu_serum
Normal              99017
Greater than 200     1485
Greater than 300     1264
Name: count, dtype: int64

In [41]:
df['A1Cresult'].value_counts()

A1Cresult
>8      92964
Norm     4990
>7       3812
Name: count, dtype: int64

In [42]:
def label_a1c_result(value):
    if pd.isnull(value) or value == 'None':
        return 'None'
    elif value == 'Norm':
        return 'Normal'
    elif value == '>7':
        return 'Greater than Seven'
    elif value == '>8':
        return 'Greater than Eight'
    else:
        return 'Unknown'
df['A1Cresult'] = df['A1Cresult'].apply(label_a1c_result)

In [43]:
df['A1Cresult'].value_counts()

A1Cresult
Greater than Eight    92964
Normal                 4990
Greater than Seven     3812
Name: count, dtype: int64

In [44]:
df['metformin'].value_counts()

metformin
No        81778
Steady    18346
Up         1067
Down        575
Name: count, dtype: int64

In [45]:
df['repaglinide'].value_counts()

repaglinide
No        100227
Steady      1384
Up           110
Down          45
Name: count, dtype: int64

In [46]:
df['nateglinide'].value_counts()

nateglinide
No        101063
Steady       668
Up            24
Down          11
Name: count, dtype: int64

In [47]:
df['chlorpropamide'].value_counts()

chlorpropamide
No        101680
Steady        79
Up             6
Down           1
Name: count, dtype: int64

In [48]:
df['glimepiride'].value_counts()

glimepiride
No        96575
Steady     4670
Up          327
Down        194
Name: count, dtype: int64

In [49]:
df['acetohexamide'].value_counts()

acetohexamide
No        101765
Steady         1
Name: count, dtype: int64

In [50]:
df['glipizide'].value_counts()

glipizide
No        89080
Steady    11356
Up          770
Down        560
Name: count, dtype: int64

In [51]:
df['tolbutamide'].value_counts()

tolbutamide
No        101743
Steady        23
Name: count, dtype: int64

In [52]:
df['pioglitazone'].value_counts()

pioglitazone
No        94438
Steady     6976
Up          234
Down        118
Name: count, dtype: int64

In [53]:
df['rosiglitazone'].value_counts()

rosiglitazone
No        95401
Steady     6100
Up          178
Down         87
Name: count, dtype: int64

In [54]:
df['acarbose'].value_counts()

acarbose
No        101458
Steady       295
Up            10
Down           3
Name: count, dtype: int64

In [55]:
df['miglitol'].value_counts()

miglitol
No        101728
Steady        31
Down           5
Up             2
Name: count, dtype: int64

In [56]:
df['troglitazone'].value_counts()

troglitazone
No        101763
Steady         3
Name: count, dtype: int64

In [57]:
df['tolazamide'].value_counts()

tolazamide
No        101727
Steady        38
Up             1
Name: count, dtype: int64

In [58]:
df['examide'].value_counts()

examide
No    101766
Name: count, dtype: int64

In [59]:
df['citoglipton'].value_counts()

citoglipton
No    101766
Name: count, dtype: int64

In [60]:
df['insulin'].value_counts()

insulin
No        47383
Steady    30849
Down      12218
Up        11316
Name: count, dtype: int64

In [61]:
df['glyburide-metformin'].value_counts()

glyburide-metformin
No        101060
Steady       692
Up             8
Down           6
Name: count, dtype: int64

In [62]:
df['glipizide-metformin'].value_counts()

glipizide-metformin
No        101753
Steady        13
Name: count, dtype: int64

In [63]:
df['glimepiride-pioglitazone'].value_counts()

glimepiride-pioglitazone
No        101765
Steady         1
Name: count, dtype: int64

In [64]:
df['metformin-rosiglitazone'].value_counts()

metformin-rosiglitazone
No        101764
Steady         2
Name: count, dtype: int64

In [65]:
df['metformin-pioglitazone'].value_counts()

metformin-pioglitazone
No        101765
Steady         1
Name: count, dtype: int64

In [66]:
df['change'].value_counts()

change
No    54755
Ch    47011
Name: count, dtype: int64

In [67]:
df['diabetesMed'].value_counts()

diabetesMed
Yes    78363
No     23403
Name: count, dtype: int64

In [68]:
df['readmitted'].value_counts()

readmitted
NO     54864
>30    35545
<30    11357
Name: count, dtype: int64

In [69]:
def classify_readmission(x):
    if x == 'NO':
        return 'Greater than 30'
    elif x == '>30':
        return 'Lesser than 30'
    elif x == '<30':
        return 'Early Readmission'
    else:
        return 'Unknown'
df['readmitted'] = df['readmitted'].apply(classify_readmission)

In [70]:
df['readmitted'].value_counts()

readmitted
Greater than 30      54864
Lesser than 30       35545
Early Readmission    11357
Name: count, dtype: int64

In [71]:
df.head(20)

encounter_id  patient_nbr             race  gender   age  weight  \
0        2278392      8222157        Caucasian  Female   5.0    87.5   
1         149190     55629189        Caucasian  Female  15.0    87.5   
2          64410     86047875  AfricanAmerican  Female  25.0    87.5   
3         500364     82442376        Caucasian    Male  35.0    87.5   
4          16680     42519267        Caucasian    Male  45.0    87.5   
5          35754     82637451        Caucasian    Male  55.0    87.5   
6          55842     84259809        Caucasian    Male  65.0    87.5   
7          63768    114882984        Caucasian    Male  75.0    87.5   
8          12522     48330783        Caucasian  Female  85.0    87.5   
9          15738     63555939        Caucasian  Female  95.0    87.5   
10         28236     89869032  AfricanAmerican  Female  45.0    87.5   
11         36900     77391171  AfricanAmerican    Male  65.0    87.5   
12         40926     85504905        Caucasian  Female  45.0    87.5   
13         42570     77586282        Caucasian    Male  85.0    87.5   
14         62256     49726791  AfricanAmerican  Female  65.0    87.5   
15         73578     86328819  AfricanAmerican    Male  65.0    87.5   
16         77076     92519352  AfricanAmerican    Male  55.0    87.5   
17         84222    108662661        Caucasian  Female  55.0    87.5   
18         89682    107389323  AfricanAmerican    Male  75.0    87.5   
19        148530     69422211        Caucasian    Male  75.0    87.5   

    admission_type_id  discharge_disposition_id  admission_source_id  \
0                   6                        25                    1   
1                   1                         1                    7   
2                   1                         1                    7   
3                   1                         1                    7   
4                   1                         1                    7   
5                   2                         1                    2   
6                   3                         1                    2   
7                   1                         1                    7   
8                   2                         1                    4   
9                   3                         3                    4   
10                  1                         1                    7   
11                  2                         1                    4   
12                  1                         3                    7   
13                  1                         6                    7   
14                  3                         1                    2   
15                  1                         3                    7   
16                  1                         1                    7   
17                  1                         1                    7   
18                  1                         1                    7   
19                  3                         6                    2   

    time_in_hospital  ... citoglipton insulin  glyburide-metformin  \
0                  1  ...          No      No                   No   
1                  3  ...          No      Up                   No   
2                  2  ...          No      No                   No   
3                  2  ...          No      Up                   No   
4                  1  ...          No  Steady                   No   
5                  3  ...          No  Steady                   No   
6                  4  ...          No  Steady                   No   
7                  5  ...          No      No                   No   
8                 13  ...          No  Steady                   No   
9                 12  ...          No  Steady                   No   
10                 9  ...          No  Steady                   No   
11                 7  ...          No  Steady                   No   
12                 7  ...          No    Down                   No  

In [72]:
df.columns

Index(['encounter_id', 'patient_nbr', 'race', 'gender', 'age', 'weight',
       'admission_type_id', 'discharge_disposition_id', 'admission_source_id',
       'time_in_hospital', 'payer_code', 'medical_specialty',
       'num_lab_procedures', 'num_procedures', 'num_medications',
       'number_outpatient', 'number_emergency', 'number_inpatient', 'diag_1',
       'diag_2', 'diag_3', 'number_diagnoses', 'max_glu_serum', 'A1Cresult',
       'metformin', 'repaglinide', 'nateglinide', 'chlorpropamide',
       'glimepiride', 'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide',
       'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',
       'tolazamide', 'examide', 'citoglipton', 'insulin',
       'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted'],
      dtype='object')

# 1. Readmission & Risk Factors

## Which variables are most predictive of early readmission (within 30 days)?

In [73]:
df['early_readmit'] = df['readmitted'].apply(lambda x: 1 if x == 'Early Readmission' else 0)
numerical_cols = ['age', 'weight', 'num_lab_procedures', 'num_procedures',
                  'num_medications', 'number_outpatient', 'number_emergency',
                  'number_inpatient', 'number_diagnoses', 'time_in_hospital']
for col in numerical_cols:
    corr, pval = pointbiserialr(df[col], df['early_readmit'])
    print(f"{col}: correlation = {corr:.3f}")

age: correlation = 0.018
weight: correlation = -0.004
num_lab_procedures: correlation = 0.020
num_procedures: correlation = -0.012
num_medications: correlation = 0.038
number_outpatient: correlation = 0.019
number_emergency: correlation = 0.061
number_inpatient: correlation = 0.165
number_diagnoses: correlation = 0.050
time_in_hospital: correlation = 0.044


### Number_inpatient: correlation = 0.165

## What percentage of diabetic patients are readmitted within 30 days?

In [74]:
early_readmitted_count = (df['readmitted'] == 'Early Readmission').sum()
total_patients = len(df)
percentage = (early_readmitted_count / total_patients) * 100
print(f"Early Readmission Rate: {percentage:.2f}%")

Early Readmission Rate: 11.16%


### Early Readmission Rate: 11.16%

### How does the number of inpatient/emergency/outpatient visits affect readmission risk?

### Number_outpatient: correlation = 0.019
### Number_emergency: correlation = 0.061
### Number_inpatient: correlation = 0.165

## Do patients with abnormal A1C or max glucose serum levels have higher readmission rates?

In [75]:
a1c_readmit = df.groupby('A1Cresult')['readmitted'].value_counts(normalize=True).unstack()
print(a1c_readmit[['Early Readmission']].sort_values('Early Readmission', ascending=False))

readmitted          Early Readmission
A1Cresult                            
Greater than Eight           0.112861
Greater than Seven           0.100472
Normal                       0.096593


In [76]:
glu_readmit = df.groupby('max_glu_serum')['readmitted'].value_counts(normalize=True).unstack()
print(glu_readmit[['Early Readmission']].sort_values('Early Readmission', ascending=False))

readmitted        Early Readmission
max_glu_serum                      
Greater than 300           0.143196
Greater than 200           0.124579
Normal                     0.111001


### Yes patients with abnormal A1C or max glucose serum levels have higher readmission rates

## What percentage of patients were readmitted within 30 days?

In [78]:
early_readmit_count = (df['readmitted'] == 'Early Readmission').sum()
total_count = len(df)
percentage = (early_readmit_count / total_count) * 100
print(f"Percentage of patients readmitted within 30 days: {percentage:.2f}%")

Percentage of patients readmitted within 30 days: 11.16%


## What features are the strongest predictors of early readmission?

## 	How does the number of prior inpatient visits correlate with readmission?

## Is there a threshold of emergency visits that significantly increases readmission likelihood?

## Are more outpatient visits associated with better or worse outcomes?

## How do A1C test results relate to early readmission?

## Are abnormal max glucose serum levels predictive of readmission?

## Do patients with more medications tend to be readmitted more often?

## Is the combination of abnormal lab values and emergency visits a strong indicator of readmission?

## Can a logistic regression model accurately classify early readmissions?

## How well do tree-based models (e.g., Random Forest) perform in predicting readmission?

## Can we build a readmission risk score to categorize patients (low, medium, high risk)?

# 2.Demographic Disparities

## Are certain races or age groups more likely to be readmitted?

In [79]:
race_readmit = df.groupby('race')['readmitted'].value_counts(normalize=True).unstack()
print(race_readmit[['Early Readmission']].sort_values('Early Readmission', ascending=False))

readmitted       Early Readmission
race                              
AfricanAmerican           0.112181
Caucasian                 0.112030
Hispanic                  0.104075
Asian                     0.101404
Other                     0.096282


In [80]:
age_readmit = df.groupby('age')['readmitted'].value_counts(normalize=True).unstack()
print(age_readmit[['Early Readmission']].sort_index())

readmitted  Early Readmission
age                          
5.0                  0.018634
15.0                 0.057887
25.0                 0.142426
35.0                 0.112318
45.0                 0.106040
55.0                 0.096662
65.0                 0.111284
75.0                 0.117731
85.0                 0.120835
95.0                 0.110992


## Do gender differences exist in length of stay or frequency of readmission?

In [81]:
df.groupby('gender')['time_in_hospital'].mean()

gender
Female             4.478797
Male               4.299777
Unknown/Invalid    3.333333
Name: time_in_hospital, dtype: float64

In [82]:
df.groupby('gender')['readmitted'].value_counts(normalize=True).unstack()[['Early Readmission']]

readmitted       Early Readmission
gender                            
Female                    0.112452
Male                      0.110615
Unknown/Invalid                NaN

## Are there disparities in the prescription of diabetes medications among different demographic groups?

In [83]:
medications = ['metformin', 'glipizide', 'glyburide', 'insulin']
for med in medications:
    print(df.groupby('race')[med].value_counts(normalize=True).unstack())

metformin            Down        No    Steady        Up
race                                                   
AfricanAmerican  0.004633  0.817855  0.166788  0.010724
Asian            0.006240  0.820593  0.168487  0.004680
Caucasian        0.005869  0.800988  0.182706  0.010437
Hispanic         0.007855  0.763868  0.214531  0.013746
Other            0.003984  0.803453  0.184595  0.007968
glipizide            Down        No    Steady        Up
race                                                   
AfricanAmerican  0.004425  0.890161  0.098386  0.007028
Asian            0.004680  0.862715  0.118565  0.014041
Caucasian        0.005729  0.872238  0.114646  0.007388
Hispanic         0.006873  0.872361  0.106529  0.014237
Other            0.005976  0.857238  0.124834  0.011952
glyburide            Down        No    Steady        Up
race                                                   
AfricanAmerican  0.003384  0.912285  0.077095  0.007236
Asian            0.003120  0.892356  0.101404  0

## Which racial groups have the highest 30-day readmission rates?

In [84]:
df.groupby('race')['readmitted'].value_counts(normalize=True).unstack()[['Early Readmission']].sort_values('Early Readmission', ascending=False)

readmitted       Early Readmission
race                              
AfricanAmerican           0.112181
Caucasian                 0.112030
Hispanic                  0.104075
Asian                     0.101404
Other                     0.096282

## Are older age groups more prone to readmission?


In [85]:
df.groupby('age')['readmitted'].value_counts(normalize=True).unstack()[['Early Readmission']]

readmitted  Early Readmission
age                          
5.0                  0.018634
15.0                 0.057887
25.0                 0.142426
35.0                 0.112318
45.0                 0.106040
55.0                 0.096662
65.0                 0.111284
75.0                 0.117731
85.0                 0.120835
95.0                 0.110992

## How do gender-based readmission rates compare?

## Are length-of-stay differences statistically significant across genders?

## Do specific race and gender combinations experience higher-than-average readmission?

## Are minorities under-prescribed A1C tests or diabetes medications?

## Do men and women differ in the number of lab procedures received?

## Are treatment changes more likely for certain demographic groups?

## Are younger patients (e.g., under 30) admitted with more severe conditions?

## Does age influence the likelihood of being discharged home vs. a facility?

## Do readmission rates vary significantly across age brackets (e.g., 50–60 vs. 70–80)?

## Are older patients more likely to receive insulin or combination drugs?

# 3. Treatment & Medication Insights

## Which diabetes medications are associated with the lowest readmission rates?

In [86]:
medications = [
    'metformin', 'glipizide', 'glyburide', 'insulin', 'pioglitazone', 
    'rosiglitazone', 'acarbose', 'nateglinide', 'repaglinide'
]
for med in medications:
    print(f"\n{med}")
    print(df.groupby(med)['readmitted'].value_counts(normalize=True).unstack()[['Early Readmission']])


metformin
readmitted  Early Readmission
metformin                    
Down                 0.120000
No                   0.115165
Steady               0.097133
Up                   0.082474

glipizide
readmitted  Early Readmission
glipizide                    
Down                 0.151786
No                   0.111192
Steady               0.111659
Up                   0.128571

glyburide
readmitted  Early Readmission
glyburide                    
Down                 0.092199
No                   0.112220
Steady               0.107289
Up                   0.104680

insulin
readmitted  Early Readmission
insulin                      
Down                 0.138975
No                   0.100374
Steady               0.111284
Up                   0.129905

pioglitazone
readmitted    Early Readmission
pioglitazone                   
Down                   0.152542
No                     0.112063
Steady                 0.104214
Up                     0.123932

rosiglitazone
readmitted     Ea

## Does changing medication during a hospital stay lead to better patient outcomes?

In [87]:
df.groupby('change')['readmitted'].value_counts(normalize=True).unstack()[['Early Readmission']]

readmitted  Early Readmission
change                       
Ch                   0.118228
No                   0.105908

## How does insulin usage correlate with hospital stay or readmission?


In [88]:
df.groupby('insulin')['time_in_hospital'].mean()

insulin
Down      4.922164
No        4.079248
Steady    4.364193
Up        5.240809
Name: time_in_hospital, dtype: float64

In [89]:
df.groupby('insulin')['readmitted'].value_counts(normalize=True).unstack()[['Early Readmission']]

readmitted  Early Readmission
insulin                      
Down                 0.138975
No                   0.100374
Steady               0.111284
Up                   0.129905

## What is the impact of combination drugs (e.g., glyburide-metformin) on patient outcomes?
 	


In [90]:
combo_drugs = ['glyburide-metformin', 'glipizide-metformin', 'glimepiride-pioglitazone', 
               'metformin-rosiglitazone', 'metformin-pioglitazone']
for drug in combo_drugs:
    print(f"\n{drug}")
    print(df.groupby(drug)['readmitted'].value_counts(normalize=True).unstack()[['Early Readmission']])


glyburide-metformin
readmitted           Early Readmission
glyburide-metformin                   
Down                          0.166667
No                            0.111607
Steady                        0.111272
Up                                 NaN

glipizide-metformin
readmitted           Early Readmission
glipizide-metformin                   
No                            0.111604
Steady                        0.076923

glimepiride-pioglitazone
readmitted                Early Readmission
glimepiride-pioglitazone                   
No                                   0.1116
Steady                                  NaN

metformin-rosiglitazone
readmitted               Early Readmission
metformin-rosiglitazone                   
No                                0.111601
Steady                                 NaN

metformin-pioglitazone
readmitted              Early Readmission
metformin-pioglitazone                   
No                                 0.1116
Steady             

## Which individual medications are associated with the lowest readmission rates?
 	

In [91]:
summary = {}
for med in medications:
    values = df.groupby(med)['readmitted'].value_counts(normalize=True).unstack()
    if 'Yes' in values.index:
        summary[med] = values.loc['Yes', 'Early Readmission']
sorted_summary = sorted(summary.items(), key=lambda x: x[1])
for med, rate in sorted_summary:
    print(f"{med}: Early Readmission Rate = {rate:.3f}")

## Do combination therapies reduce readmission more effectively than single drugs?
 	

## Are insulin users more likely to be readmitted than non-insulin users?
 	

## Does metformin use correlate with shorter hospital stays?
 	

## Are there medications frequently changed during the stay that reduce risk?
 	

## Do patients who received "No" for diabetes medication have worse outcomes?
 	

## Which medications are more often associated with abnormal A1C values?
 	

## What is the average number of medications for readmitted vs. non-readmitted patients?
 	

## Are certain medications only prescribed to certain demographics?
 	

## What medication patterns are most common in non-readmitted patients?
 	

## Is there an interaction between medication type and number of diagnoses?
 	

## Do patients who experienced a medication change have better A1C improvements?

# 4. Hospital Performance & Process Analysis

## Which admission types (e.g., emergency vs. elective) are associated with the highest readmission rates?

In [92]:
df.groupby('admission_type_id')['readmitted'].value_counts(normalize=True).unstack()[['Early Readmission']]

readmitted         Early Readmission
admission_type_id                   
1                           0.115225
2                           0.111797
3                           0.103927
4                           0.100000
5                           0.103448
6                           0.110754
7                                NaN
8                           0.084375

In [93]:
admission_type_map = {
    1: 'Emergency', 2: 'Urgent', 3: 'Elective', 4: 'Newborn',
    5: 'Not Available', 6: 'NULL', 7: 'Trauma Center', 8: 'Not Mapped'
}
df['admission_type'] = df['admission_type_id'].map(admission_type_map)

In [94]:
df['admission_type'].value_counts()

admission_type
Emergency        53990
Elective         18869
Urgent           18480
NULL              5291
Not Available     4785
Not Mapped         320
Trauma Center       21
Newborn             10
Name: count, dtype: int64

## How do discharge dispositions affect the likelihood of readmission?

In [95]:
df.groupby('discharge_disposition_id')['readmitted'].value_counts(normalize=True).unstack()[['Early Readmission']]

readmitted                Early Readmission
discharge_disposition_id                   
1                                  0.093004
2                                  0.160714
3                                  0.146625
4                                  0.127607
5                                  0.208615
6                                  0.126957
7                                  0.144462
8                                  0.138889
9                                  0.428571
10                                      NaN
11                                      NaN
12                                 0.666667
13                                 0.047619
14                                 0.064516
15                                 0.444444
16                                      NaN
17                                      NaN
18                                 0.124357
19                                      NaN
20                                      NaN
22                                 0.276969
23                                 0.072816
24                                 0.145833
25                                 0.093023
27                                      NaN
28                                 0.366906

## Which medical specialties are most effective at preventing readmissions?

In [96]:
df.groupby('medical_specialty')['readmitted'].value_counts(normalize=True).unstack()[['Early Readmission']].sort_values('Early Readmission')

readmitted                        Early Readmission
medical_specialty                                  
Pediatrics-Endocrinology                   0.006289
Otolaryngology                             0.016000
Pediatrics-CriticalCare                    0.034483
ObstetricsandGynecology                    0.047690
Endocrinology                              0.050000
...                                             ...
Psychiatry-Addictive                            NaN
Speech                                          NaN
SportsMedicine                                  NaN
Surgery-Pediatric                               NaN
Surgery-PlasticwithinHeadandNeck                NaN

[72 rows x 1 columns]

In [97]:
specialty_counts = df['medical_specialty'].value_counts()
top_specialties = specialty_counts[specialty_counts > 100]  # threshold
df_filtered = df[df['medical_specialty'].isin(top_specialties.index)]

## Is there a correlation between the number of lab procedures and better health outcomes?

In [127]:
from scipy.stats import pointbiserialr
corr, pval = pointbiserialr(df['num_lab_procedures'], df['readmitted'] == 'Early Readmission')
print(f"Correlation: {corr:.3f}, p-value: {pval:.3f}")

Correlation: 0.020, p-value: 0.000


8.207771005578811e-11

### Positive correlation may suggest sicker patients getting more labs

## Which admission types have the highest readmission rates?
 	

In [99]:
admission_readmit = df.groupby('admission_type')['readmitted'].value_counts(normalize=True).unstack()
admission_readmit

readmitted      Early Readmission  Greater than 30  Lesser than 30
admission_type                                                    
Elective                 0.103927         0.591552        0.304521
Emergency                0.115225         0.527135        0.357640
NULL                     0.110754         0.467586        0.421659
Newborn                  0.100000         0.700000        0.200000
Not Available            0.103448         0.536886        0.359666
Not Mapped               0.084375         0.653125        0.262500
Trauma Center                 NaN         1.000000             NaN
Urgent                   0.111797         0.539069        0.349134

## Are elective admissions less likely to be readmitted than emergency ones?
 	

## How does admission source (e.g., referral vs. emergency room) affect readmission?
 	

## What discharge disposition types correlate with highest readmission risk?
 	

## Do patients discharged to skilled nursing facilities have longer stays?
 	

## Which medical specialties are associated with the lowest readmissions?
 	

## Are certain specialties more aggressive in ordering lab tests?
 	

## Does medical specialty influence length of stay or medication usage?
 	

## Are patients in general practice less likely to receive A1C testing?
 	

## Which discharge categories (e.g., expired, home, transfer) relate to patient mortality?
 	

## Does frequency of lab procedures reflect better outcomes?
 	

## Can we rank hospitals or departments based on performance using readmission and length of stay?

# 5. Length of Stay Analysis

## What are the top predictors of longer hospital stays for diabetic patients? 	

In [100]:
numeric_cols = ['age', 'weight', 'num_lab_procedures', 'num_procedures',
                'num_medications', 'number_outpatient', 'number_emergency',
                'number_inpatient', 'number_diagnoses']
for col in numeric_cols:
    corr = df[col].corr(df['time_in_hospital'])
    print(f"{col}: correlation = {corr:.3f}")

age: correlation = 0.108
weight: correlation = -0.007
num_lab_procedures: correlation = 0.318
num_procedures: correlation = 0.191
num_medications: correlation = 0.466
number_outpatient: correlation = -0.009
number_emergency: correlation = -0.010
number_inpatient: correlation = 0.074
number_diagnoses: correlation = 0.220


## Is there a relationship between length of stay and readmission? 	

In [101]:
df.groupby('readmitted')['time_in_hospital'].mean()

readmitted
Early Readmission    4.768249
Greater than 30      4.254429
Lesser than 30       4.495541
Name: time_in_hospital, dtype: float64

## Do certain diagnoses (diag_1/2/3) lead to significantly longer hospitalizations?

In [102]:
top_diag = df['diag_1'].value_counts().head(10).index
df[df['diag_1'].isin(top_diag)].groupby('diag_1')['time_in_hospital'].mean().sort_values(ascending=False)

diag_1
486    5.035063
410    4.874377
682    4.807542
434    4.761341
491    4.753407
428    4.572280
715    3.782892
427    3.600868
414    3.471661
786    2.121763
Name: time_in_hospital, dtype: float64

## What is the distribution of hospital stays for diabetic patients?

In [104]:
df['time_in_hospital'].describe()

count    101766.000000
mean          4.395987
std           2.985108
min           1.000000
25%           2.000000
50%           4.000000
75%           6.000000
max          14.000000
Name: time_in_hospital, dtype: float64

## What variables most strongly influence the length of hospital stay? 	

In [105]:
df.groupby('A1Cresult')['time_in_hospital'].mean()
df.groupby('max_glu_serum')['time_in_hospital'].mean()
df.groupby('insulin')['time_in_hospital'].mean()
df.groupby('change')['time_in_hospital'].mean()
df.groupby('gender')['time_in_hospital'].mean()
df.groupby('race')['time_in_hospital'].mean()

race
AfricanAmerican    4.507860
Asian              3.995320
Caucasian          4.382930
Hispanic           4.059892
Other              4.273572
Name: time_in_hospital, dtype: float64

## How does time in hospital correlate with readmission risk?

## Do more lab procedures increase or decrease length of stay?

## How does length of stay differ across medical specialties? 	

## Are longer stays associated with better A1C or glucose control? 	

## Is there a point beyond which a longer stay no longer reduces readmission? 	

## What diagnoses are most associated with prolonged hospitalization? 	

## Do certain medications correlate with shorter stays? 	

## Are readmitted patients typically those who stayed shorter or longer initially? 	

## What is the average length of stay by admission type? 	

## Does a longer stay predict higher number of procedures or tests?

# 6. Diagnosis & Comorbidity Insights

## What are the most common comorbid diagnoses among diabetic patients?

In [108]:
from collections import Counter
comorbid = pd.concat([df['diag_2'], df['diag_3']])
comorbid_counts = comorbid.value_counts()
comorbid_counts

250    19049
276    12285
401    12025
428    11239
427     8991
       ...  
947        1
520        1
V69        1
974        1
971        1
Name: count, Length: 848, dtype: int64

In [109]:
comorbid_counts = comorbid.value_counts().head(10)
comorbid_counts

250    19049
276    12285
401    12025
428    11239
427     8991
414     6314
496     5910
599     5229
403     5180
585     3863
Name: count, dtype: int64

## How does the number of diagnoses relate to readmission or mortality risk?

In [110]:
df.groupby('readmitted')['number_diagnoses'].mean()

readmitted
Early Readmission    7.692789
Greater than 30      7.221366
Lesser than 30       7.646898
Name: number_diagnoses, dtype: float64

## Are there certain diagnosis patterns that indicate high-risk diabetic patients?

In [112]:
early = df[df['readmitted'] == 'Early Readmission']
combo_counts = early.groupby(['diag_1', 'diag_2', 'diag_3']).size().sort_values(ascending=False).head(10)
combo_counts

diag_1  diag_2  diag_3
414     411     250       23
715     401     250       23
410     414     250       20
414     411     401       19
                428       19
434     250     401       18
414     411     496       18
996     403     585       16
414     411     427       16
715     250     401       16
dtype: int64

## What are the top 10 most frequent primary diagnoses (diag_1)?

In [113]:
df['diag_1'].value_counts().head(10)

diag_1
428    6883
414    6581
786    4016
410    3614
486    3508
427    2766
491    2275
715    2151
682    2042
434    2028
Name: count, dtype: int64

## Which combinations of diag_1, diag_2, diag_3 are most common?

In [114]:
common_combos = df.groupby(['diag_1', 'diag_2', 'diag_3']).size().sort_values(ascending=False).head(10)
common_combos

diag_1  diag_2  diag_3
414     411     250       350
                401       271
786     401     250       247
410     414     250       235
786     250     401       231
715     401     250       208
414     401     250       197
715     250     401       188
414     250     401       187
        411     428       172
dtype: int64

## Which diagnosis combinations are associated with the highest readmission rates?

## Do comorbidities like hypertension or kidney disease influence readmission?

## How does the number of diagnoses relate to time in hospital?

## What are the most common comorbid patterns among patients who died?

## Are certain diagnoses more common in younger vs. older patients?

## How do patients with 9+ diagnoses differ in outcomes from those with <3?

## Which diagnostic categories correlate most with abnormal glucose levels?

## Do patients with heart-related diagnoses stay longer than others?

## Can we predict likely readmission based solely on diagnosis codes?

## Are certain diagnosis combinations tied to higher medication counts?

# 7. Utilization & Resource Efficiency

## How does the number of medications or lab procedures relate to outcomes?
 	

In [115]:
df.groupby('readmitted')[['num_medications', 'num_lab_procedures']].mean()

num_medications  num_lab_procedures
readmitted                                            
Early Readmission        16.903143           44.226028
Greater than 30          15.670367           42.381598
Lesser than 30           16.282768           43.836601

## Are more procedures always linked to better results, or is there a point of diminishing returns?
 	

In [117]:
df['procedure_bin'] = pd.cut(df['num_procedures'], bins=[-1, 0, 1, 2, 3, 4, 5, 10, 20], right=True)
early_readmit_rate = df[df['readmitted'] == 'Early Readmission'].groupby('procedure_bin').size() / df.groupby('procedure_bin').size()
early_readmit_rate

C:\Users\DLK Groups\AppData\Local\Temp\ipykernel_6872\3542623539.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  early_readmit_rate = df[df['readmitted'] == 'Early Readmission'].groupby('procedure_bin').size() / df.groupby('procedure_bin').size()
C:\Users\DLK Groups\AppData\Local\Temp\ipykernel_6872\3542623539.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  early_readmit_rate = df[df['readmitted'] == 'Early Readmission'].groupby('procedure_bin').size() / df.groupby('procedure_bin').size()


procedure_bin
(-1, 0]     0.110778
(0, 1]      0.122071
(1, 2]      0.111819
(2, 3]      0.106852
(3, 4]      0.110287
(4, 5]      0.094217
(5, 10]     0.095882
(10, 20]         NaN
dtype: float64

## What is the average resource utilization (labs, procedures) for patients who are not readmitted?
 	


In [119]:
df[df['readmitted'] == 'Greater than 30'][['num_lab_procedures', 'num_procedures', 'num_medications']].describe()

num_lab_procedures  num_procedures  num_medications
count        54864.000000    54864.000000     54864.000000
mean            42.381598        1.410305        15.670367
std             19.796262        1.739693         8.427628
min              1.000000        0.000000         1.000000
25%             30.000000        0.000000        10.000000
50%             44.000000        1.000000        14.000000
75%             56.000000        2.000000        20.000000
max            126.000000        6.000000        79.000000

## How does the number of lab procedures relate to readmission risk?
 	

In [120]:
from scipy.stats import pointbiserialr
df['readmit_binary'] = df['readmitted'].apply(lambda x: 1 if x == 'Early Readmission' else 0)
corr, pval = pointbiserialr(df['num_lab_procedures'], df['readmit_binary'])
print(f"Correlation = {corr:.3f}, p-value = {pval:.3f}")

Correlation = 0.020, p-value = 0.000


## Do patients with more medications always fare better?
 	

In [121]:
df.groupby('readmitted')['num_medications'].mean()

readmitted
Early Readmission    16.903143
Greater than 30      15.670367
Lesser than 30       16.282768
Name: num_medications, dtype: float64

## Is there a diminishing return effect in number of procedures and outcomes?
 	

## What is the average number of lab procedures for non-readmitted patients?
 	

## How does resource usage vary by age or gender?
 	

## Are high-utilization patients receiving better care, or just more interventions?
 	

## Do more procedures correlate with longer stays or fewer readmissions?
 	

## Is there a consistent pattern in lab tests for patients with normal A1C results?
 	

## Do patients with the same diagnosis receive varying levels of care?
 	

## Are there outliers with very high resource use and poor outcomes?
 	

## Is there a “sweet spot” of lab tests/medications that minimizes readmission?
 	

## Can we cluster patients based on utilization patterns and outcome?

# 8. Temporal Patterns 

## How have readmission rates changed over the 1999–2008 period?

In [122]:
df['year'] = df['encounter_id'].astype(str).str[:4].astype(int)  # Assuming year is encoded in the ID
readmit_rate = df.groupby('year')['readmitted'].apply(lambda x: (x == 'Early Readmission').mean())
readmit_rate

year
1000    0.074074
1001    0.081081
1002    0.085714
1003    0.095238
1004    0.096774
          ...   
9995    0.000000
9996    0.000000
9997    0.000000
9998    0.125000
9999    0.000000
Name: readmitted, Length: 8385, dtype: float64


## Has the use of specific medications (e.g., metformin) increased or decreased over time?

In [123]:
df['metformin_use'] = df['metformin'].isin(['Up', 'Steady'])
metformin_trend = df.groupby('year')['metformin_use'].mean()
print(metformin_trend)


year
1000    0.129630
1001    0.243243
1002    0.257143
1003    0.142857
1004    0.064516
          ...   
9995    0.000000
9996    0.000000
9997    0.500000
9998    0.125000
9999    0.250000
Name: metformin_use, Length: 8385, dtype: float64


## Are there improvements in treatment protocols over the 10-year span?

In [124]:
df.groupby('year')['change'].value_counts(normalize=True).unstack().fillna(0)

change        Ch        No
year                      
1000    0.314815  0.685185
1001    0.486486  0.513514
1002    0.457143  0.542857
1003    0.380952  0.619048
1004    0.322581  0.677419
...          ...       ...
9995    0.000000  1.000000
9996    0.000000  1.000000
9997    0.500000  0.500000
9998    0.375000  0.625000
9999    0.000000  1.000000

[8385 rows x 2 columns]

## How did readmission rates trend from 1999 to 2008?

In [125]:
df['year'] = df['encounter_id'].astype(str).str[:4].astype(int)  # Assuming year is encoded in the ID
readmit_rate = df.groupby('year')['readmitted'].apply(lambda x: (x == 'Early Readmission').mean())
readmit_rate


year
1000    0.074074
1001    0.081081
1002    0.085714
1003    0.095238
1004    0.096774
          ...   
9995    0.000000
9996    0.000000
9997    0.000000
9998    0.125000
9999    0.000000
Name: readmitted, Length: 8385, dtype: float64


## Was there an increase in average length of stay over time?

In [126]:
los_trend = df.groupby('year')['time_in_hospital'].mean()
los_trend

year
1000    4.314815
1001    4.108108
1002    4.200000
1003    4.952381
1004    4.016129
          ...   
9995    4.000000
9996    8.000000
9997    5.500000
9998    3.375000
9999    3.250000
Name: time_in_hospital, Length: 8385, dtype: float64


## Did metformin or insulin usage increase or decrease over the years?

## Are newer drugs (e.g., rosiglitazone) more common in later years?

## Were there specific years with spikes in lab test frequency?

## Has the number of diagnoses per patient increased over the years?

## Did outcomes improve as treatment protocols evolved over time?

## Were treatment changes more frequent in early or later years?

## Were earlier years associated with higher mortality or longer stays?

## How did the use of combination therapies change over time?

## Was there a shift in dominant diagnosis codes across years?

## Are more recent records associated with lower readmission and better A1C control?